src/net.cpp

```diff
@@ -1119,32 +1157,16 @@ void CConnman::DisconnectNodes()
         {
             if (pnode->fDisconnect)
             {
-                // remove from m_nodes
-                m_nodes.erase(remove(m_nodes.begin(), m_nodes.end(), pnode), m_nodes.end());
-
-                // release outbound grant (if any)
-                pnode->grantOutbound.Release();
-
-                // close socket and cleanup
-                pnode->CloseSocketDisconnect();
-
-                // hold in disconnected pool until all refs are released
-                pnode->Release();
-                m_nodes_disconnected.push_back(pnode);
+                DisconnectAndReleaseNode(pnode);
             }
         }
     }
+    // Delete disconnected nodes
+    std::list<CNode*> disconnected_nodes_copy{};
+    WITH_LOCK(m_nodes_disconnected_mutex, disconnected_nodes_copy = m_nodes_disconnected);
+    for (CNode* pnode : disconnected_nodes_copy)
     {
-        // Delete disconnected nodes
-        std::list<CNode*> nodes_disconnected_copy = m_nodes_disconnected;
-        for (CNode* pnode : nodes_disconnected_copy)
-        {
-            // Destroy the object only after other threads have stopped using it.
-            if (pnode->GetRefCount() <= 0) {
-                m_nodes_disconnected.remove(pnode);
-                DeleteNode(pnode);
-            }
-        }
+        DeleteDisconnectedNode(pnode);
     }
```

Comment:
Right, this is what I tried to prevent against in an [earlier version](https://github.com/bitcoin/bitcoin/commit/7ad6edb44d599bdd9f8c6e5aae5899c97436b675#diff-00021eed586a482abdb09d6cdada1d90115abe988a91421851960e26658bed02R1152-R1154), but in that version I think it was [superfluous](https://github.com/bitcoin/bitcoin/pull/27912#discussion_r1233973528), however here it would be appropriate to use so that only one thread would actually perform the deletion ever.

Reply:
`DeleteNode()` now gated behind `if (pnode->nRefCount.compare_exchange_strong(expectedRefCount, -1))` so it can only be called once (by a single thread).

Reply:
I love this lock free pattern!

The variable used as a guard (`nRefCount`) has to be outside of the object being protected. Otherwise if two threads execute it concurrently the first one may swap from 0 to -1 and `delete` the object. Then the second thread will end up reading freed memory when it tries to use `nRefCount`.

This should be safe:

```cpp
    // Delete disconnected nodes. Call DeleteNode() without holding m_nodes_mutex or m_nodes_disconnected_mutex.
    std::vector<CNode*> to_delete;
    { 
        LOCK(m_nodes_disconnected_mutex);
        for (auto it = m_nodes_disconnected.begin(); it != m_nodes_disconnected.end();) { 
            CNode* node = *it;
            if (node->GetRefCount() == 0) { 
                it = m_nodes_disconnected.erase(it);
                to_delete.push_back(node);
            } else { 
                ++it;
            } 
        } 
    } 
    for (CNode* node : to_delete) { 
        DeleteNode(node);
    } 
```